### DISCUSSION TITLE
Magick++ encodes to PNG sequentially even with multiple threads

### DISCUSSION THREAD

**Discussion by atk91** [ImageMagick/ImageMagick#4123] [Help]
Version: ImageMagick 7.1.0-4 Q16 x86_64 2021-07-18

Source image:
![stark](https://user-images.githubusercontent.com/3193829/131025779-2199a70b-08c7-4be9-ac95-979d57c2b846.jpg)

Compile the code with 
```
g++ `Magick++-config --cxxflags --cppflags` -O2 -o demo main.cpp `Magick++-config --ldflags --libs `
```

```
#include <Magick++.h> 
#include <iostream> 
#include <vector>
#include <thread>
#include <mutex>

using namespace std; 
using namespace Magick; 


void log(const char* message) {
    static std::mutex log_mutex;
    std::lock_guard<std::mutex> guard(log_mutex);
    std::cout << message << "\n";
}

Blob doStuff() {
    Image image;
    image.read( "stark.jpg" );
    log("Starting scale");
    image.scale("3000x3000!");
    log("Scale done");
    Blob res;
    log("Starting write");
    image.write(&res, "png");
    log("Write done");
    return res;
}

int main(int argc,char **argv) 
{ 
  InitializeMagick(*argv);

  std::vector<std::thread> ts;
  for (int i = 0; i < 8; i++) {
      ts.emplace_back(doStuff);
  }
  for (std::thread& t: ts) {
      t.join();
  }
  std::cout << "All done";
  return 0; 
} 
```

As you can see from logs, once that scaling is done, the `write`ing is done sequentially! You can even see from `top` that only one cpu is being utilized after all threads print `Starting write`. It seems that PNG encoder has mutex somewhere inside which prevents encoding to png in multiple threads. Why is it so and how to encode images in parallel?

There's no such problem with `webp`, for example.

**Comment by snibgo** [comment]
I don't understand your multi-threading, so I can't answer your question, but it looks as if each thread is attempting to write to the same file. Is that correct?

**Reply by atk91** [reply]
Not exactly.
Each thread writes to its own `Blob`, which is declared in `doStuff`.

**Reply by snibgo** [reply]
Yes, my mistake, sorry, I misread the "write".

**Comment by atk91** [comment]
I opened an issue https://github.com/ImageMagick/ImageMagick/issues/4134

I might be wrong, however I think that that lock is being hold for absolutely no reason.

**Comment by dlemstra** [comment]
I don't know how you compiled ImageMagick but it looks like `IMPNG_SETJMP_NOT_THREAD_SAFE` is defined in your situation. And that is happening because `IMPNG_SETJMP_IS_THREAD_SAFE` is not defined.

**Reply by atk91** [reply]
I've just `pacman -S imagemagick`. Thanks for the clue, I'll try to find out why Arch and Alpine maintainers do not define this.

**Reply by atk91** [reply]
@dlemstra hmm, [here](https://github.com/dlemstra/Magick.NET/discussions/623) you wrote that it's now being set.
Arch Linux build is not explicit about that https://github.com/archlinux/svntogit-packages/blob/packages/imagemagick/trunk/PKGBUILD

**Reply by dlemstra** [reply]
It looks like this was still not configured properly, thanks for pointing me to that issue. The png coder checked for `IMPNG_SETJMP_IS_THREAD_SAFE`. On Windows this define was set. But the define was set to `SETJMP_IS_THREAD_SAFE` in `configure.ac` that is being used by the other platforms. Just pushed a patch to resolve this in the next release.

**Reply by atk91** [reply]
@dlemstra It still does not getting set somewhy.

I added
`fprintf(stdout, "line %d: IMPNG_SETJMP_NOT_THREAD_SAFE\n", __LINE__);`
just after
`#ifdef IMPNG_SETJMP_NOT_THREAD_SAFE` in coders/png.c on line 8105 on current main branch, which is 7e36bce9e429604a633b921b01d26ebab00e5578 at the moment

compiled with `./configure && make`
and then when I do `./utilities/magick convert /tmp/img.jpg /tmp/img.png` I get

`line 8106: IMPNG_SETJMP_NOT_THREAD_SAFE` printed to stdout

**Reply by atk91** [reply]
`#include "config/config.h"` in `coders/png.c` seems to do the trick, however I'm not sure whether that's the right thing to do

**Reply by atk91** [reply]
@dlemstra is this correct thing to do? I may do a PR

**Reply by urban-warrior** [reply]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow.
